File: sdk/typescript/src/transport.ts
/**
 * Transport layer — stdio and WebSocket implementations.
 *
 * Both transports speak JSONL (one JSON-RPC message per line).
 */

import { spawn, type ChildProcess } from 'child_process'
import { createInterface } from 'readline'
import type { JsonRpcRequest, JsonRpcResponse, JsonRpcNotification } from './types.js'

export interface Transport {
  send(request: JsonRpcRequest): Promise<JsonRpcResponse>
  onNotification(callback: (n: JsonRpcNotification) => void): void
  close(): Promise<void>
  isConnected(): boolean
}

type NotificationCallback = (n: JsonRpcNotification) => void

/**
 * Stdio transport — spawns `legnacode app-server --transport stdio`
 * and communicates via stdin/stdout JSONL.
 */
export class StdioTransport implements Transport {
  private proc: ChildProcess | null = null
  private pending = new Map<string | number, {
    resolve: (r: JsonRpcResponse) => void
    reject: (e: Error) => void
  }>()
  private notificationListeners: NotificationCallback[] = []
  private connected = false

  constructor(
    private binaryPath: string = 'legnacode',
    private workingDir?: string,
    private env?: Record<string, string>,
  ) {}

  /** Spawn the subprocess and start listening. */
  async start(): Promise<void> {
    if (this.connected) return

    this.proc = spawn(this.binaryPath, ['app-server', '--transport', 'stdio'], {
      cwd: this.workingDir,
      env: { ...process.env, ...this.env },
      stdio: ['pipe', 'pipe', 'pipe'],
    })

    const rl = createInterface({ input: this.proc.stdout!, terminal: false })
    rl.on('line', (line) => {
      const trimmed = line.trim()
      if (!trimmed) return
      try {
        const msg = JSON.parse(trimmed)
        if ('id' in msg && (msg.result !== undefined || msg.error !== undefined)) {
          const pending = this.pending.get(msg.id)
          if (pending) {
            this.pending.delete(msg.id)
            pending.resolve(msg as JsonRpcResponse)
          }
        } else if ('method' in msg && !('id' in msg)) {
          for (const cb of this.notificationListeners) cb(msg as JsonRpcNotification)
        }
      } catch { /* ignore malformed lines */ }
    })

    this.proc.on('exit', () => {
      this.connected = false
      for (const [, p] of this.pending) p.reject(new Error('Process exited'))
      this.pending.clear()
    })

    this.connected = true
  }

  async send(request: JsonRpcRequest): Promise<JsonRpcResponse> {
    if (!this.proc?.stdin?.writable) throw new Error('Transport not connected')
    return new Promise((resolve, reject) => {
      this.pending.set(request.id, { resolve, reject })
      this.proc!.stdin!.write(JSON.stringify(request) + '\n')
    })
  }

  onNotification(callback: NotificationCallback): void {
    this.notificationListeners.push(callback)
  }

  async close(): Promise<void> {
    this.connected = false
    this.proc?.kill('SIGTERM')
    this.proc = null
    for (const [, p] of this.pending) p.reject(new Error('Transport closed'))
    this.pending.clear()
  }

  isConnected(): boolean {
    return this.connected
  }
}

/**
 * WebSocket transport — connects to a running legnacode app-server
 * WebSocket endpoint.
 */
export class WebSocketTransport implements Transport {
  private ws: import('ws').WebSocket | null = null
  private pending = new Map<string | number, {
    resolve: (r: JsonRpcResponse) => void
    reject: (e: Error) => void
  }>()
  private notificationListeners: NotificationCallback[] = []
  private connected = false

  constructor(
    private url: string = 'ws://127.0.0.1:3100',
  ) {}

  async start(): Promise<void> {
    if (this.connected) return
    const { default: WebSocket } = await import('ws')
    this.ws = new WebSocket(this.url)

    await new Promise<void>((resolve, reject) => {
      this.ws!.on('open', () => { this.connected = true; resolve() })
      this.ws!.on('error', reject)
    })

    this.ws.on('message', (data) => {
      try {
        const msg = JSON.parse(String(data))
        if ('id' in msg && (msg.result !== undefined || msg.error !== undefined)) {
          const pending = this.pending.get(msg.id)
          if (pending) { this.pending.delete(msg.id); pending.resolve(msg) }
        } else if ('method' in msg && !('id' in msg)) {
          for (const cb of this.notificationListeners) cb(msg)
        }
      } catch { /* ignore */ }
    })

    this.ws.on('close', () => {
      this.connected = false
      for (const [, p] of this.pending) p.reject(new Error('WebSocket closed'))
      this.pending.clear()
    })
  }

  async send(request: JsonRpcRequest): Promise<JsonRpcResponse> {
    if (!this.ws || !this.connected) throw new Error('Transport not connected')
    return new Promise((resolve, reject) => {
      this.pending.set(request.id, { resolve, reject })
      this.ws!.send(JSON.stringify(request))
    })
  }

  onNotification(callback: NotificationCallback): void {
    this.notificationListeners.push(callback)
  }

  async close(): Promise<void> {
    this.connected = false
    this.ws?.close()
    this.ws = null
    for (const [, p] of this.pending) p.reject(new Error('Transport closed'))
    this.pending.clear()
  }

  isConnected(): boolean {
    return this.connected
  }
}

File: sdk/python/legnacode/transport.py
"""
Transport layer — stdio and WebSocket implementations.

Both transports speak JSONL (one JSON-RPC message per line).
"""

from __future__ import annotations

import asyncio
import json
import os
import subprocess
from abc import ABC, abstractmethod
from typing import Any, Callable, Optional

NotificationCallback = Callable[[dict[str, Any]], None]


class Transport(ABC):
    """Abstract transport interface."""

    @abstractmethod
    async def send(self, request: dict[str, Any]) -> dict[str, Any]: ...

    @abstractmethod
    def on_notification(self, callback: NotificationCallback) -> None: ...

    @abstractmethod
    async def close(self) -> None: ...

    @abstractmethod
    def is_connected(self) -> bool: ...


class StdioTransport(Transport):
    """
    Stdio transport — spawns `legnacode app-server --transport stdio`
    and communicates via stdin/stdout JSONL.
    """

    def __init__(
        self,
        binary_path: str = "legnacode",
        working_dir: Optional[str] = None,
        env: Optional[dict[str, str]] = None,
    ):
        self._binary = binary_path
        self._cwd = working_dir
        self._env = {**os.environ, **(env or {})}
        self._proc: Optional[asyncio.subprocess.Process] = None
        self._pending: dict[int | str, asyncio.Future[dict[str, Any]]] = {}
        self._listeners: list[NotificationCallback] = []
        self._connected = False
        self._reader_task: Optional[asyncio.Task[None]] = None
    async def start(self) -> None:
        if self._connected:
            return
        self._proc = await asyncio.create_subprocess_exec(
            self._binary, "app-server", "--transport", "stdio",
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            cwd=self._cwd,
            env=self._env,
        )
        self._connected = True
        self._reader_task = asyncio.create_task(self._read_loop())

    async def _read_loop(self) -> None:
        assert self._proc and self._proc.stdout
        while self._connected:
            line = await self._proc.stdout.readline()
            if not line:
                self._connected = False
                break
            try:
                msg = json.loads(line.decode().strip())
            except (json.JSONDecodeError, UnicodeDecodeError):
                continue
            if "id" in msg and ("result" in msg or "error" in msg):
                fut = self._pending.pop(msg["id"], None)
                if fut and not fut.done():
                    fut.set_result(msg)
            elif "method" in msg and "id" not in msg:
                for cb in self._listeners:
                    cb(msg)

    async def send(self, request: dict[str, Any]) -> dict[str, Any]:
        if not self._proc or not self._proc.stdin:
            raise RuntimeError("Transport not connected")
        loop = asyncio.get_running_loop()
        fut: asyncio.Future[dict[str, Any]] = loop.create_future()
        self._pending[request["id"]] = fut
        self._proc.stdin.write((json.dumps(request) + "\n").encode())
        await self._proc.stdin.drain()
        return await fut

    def on_notification(self, callback: NotificationCallback) -> None:
        self._listeners.append(callback)

    async def close(self) -> None:
        self._connected = False
        if self._reader_task:
            self._reader_task.cancel()
        if self._proc:
            self._proc.terminate()
            await self._proc.wait()
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(RuntimeError("Transport closed"))
        self._pending.clear()

    def is_connected(self) -> bool:
        return self._connected


class WebSocketTransport(Transport):
    """
    WebSocket transport — connects to a running legnacode app-server
    WebSocket endpoint.
    """

    def __init__(self, url: str = "ws://127.0.0.1:3100"):
        self._url = url
        self._ws: Any = None
        self._pending: dict[int | str, asyncio.Future[dict[str, Any]]] = {}
        self._listeners: list[NotificationCallback] = []
        self._connected = False
        self._reader_task: Optional[asyncio.Task[None]] = None

    async def start(self) -> None:
        if self._connected:
            return
        try:
            import websockets
        except ImportError:
            raise ImportError("Install websockets: pip install legnacode-sdk[websocket]")
        self._ws = await websockets.connect(self._url)
        self._connected = True
        self._reader_task = asyncio.create_task(self._read_loop())

    async def _read_loop(self) -> None:
        try:
            async for raw in self._ws:
                try:
                    msg = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                if "id" in msg and ("result" in msg or "error" in msg):
                    fut = self._pending.pop(msg["id"], None)
                    if fut and not fut.done():
                        fut.set_result(msg)
                elif "method" in msg and "id" not in msg:
                    for cb in self._listeners:
                        cb(msg)
        except Exception:
            self._connected = False

    async def send(self, request: dict[str, Any]) -> dict[str, Any]:
        if not self._ws:
            raise RuntimeError("Transport not connected")
        loop = asyncio.get_running_loop()
        fut: asyncio.Future[dict[str, Any]] = loop.create_future()
        self._pending[request["id"]] = fut
        await self._ws.send(json.dumps(request))
        return await fut

    def on_notification(self, callback: NotificationCallback) -> None:
        self._listeners.append(callback)

    async def close(self) -> None:
        self._connected = False
        if self._reader_task:
            self._reader_task.cancel()
        if self._ws:
            await self._ws.close()
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(RuntimeError("Transport closed"))
        self._pending.clear()

    def is_connected(self) -> bool:
        return self._connected
